### ISSUE TITLE
Running out of file descriptors when reusing stream sockets for connections

### ISSUE THREAD

**Issue by xraw** [gramineproject/gramine#326]
## Description of the problem
While performing periodically hostname lookups with ```getaddrinfo```, gramine seems to reuse stream
sockets for connections to the dns server.

The following C program calls ```getaddrinfo``` in a loop in order to lookup an IP address for a host.
While running and observing with ```ss -unp``` for connections, a steady growth of sockets occurs which
are not cleaned up. Sooner or later, the process reaches the ulimit and fails with further operations.

While checking the gramine logs and the output of ss there is a correlation of gramines log
_"reconnect on a stream socket"_ with the number of sockets reported by ss.

Having a look in [gramine/LibOS/shim/src/sys/shim_socket.c](https://github.com/gramineproject/gramine/blob/master/LibOS/shim/src/sys/shim_socket.c) -> [shim_do_connect(...)](https://github.com/gramineproject/gramine/blob/66881cca331402825cf1b5f8c4f949a2c758892b/LibOS/shim/src/sys/shim_socket.c#L67) (I have added the
exclamation comments) :

```
    if (state == SOCK_CONNECTED) {
        if (addr->sa_family == AF_UNSPEC) {
            // !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
            sock->sock_state = SOCK_CREATED;
            if (sock->sock_type == SOCK_STREAM && hdl->pal_handle) {
            // !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
                DkStreamDelete(hdl->pal_handle, PAL_DELETE_ALL); // TODO: handle errors
                DkObjectClose(hdl->pal_handle);
                hdl->pal_handle = NULL;
                pal_handle_updated = true;
            }
            log_debug("shim_connect: reconnect on a stream socket");
            ret = 0;
            goto out;
        }

        log_debug("shim_connect: reconnect on a stream socket");
        ret = -EISCONN;
        goto out;
    }
```

I modified it slightly, according to my understanding of the code, fixes the resource leak. I am not sure
that this is the right approach, logically yes, but maybe there is a better way for the reconnection handling.
Furthermore, I made sure that 'getaddrinfo' gets a hints struct with AF_INET in order to bypass the AF_UNSPEC
case but this is not under my control.

```
    if (state == SOCK_CONNECTED) {
        if (addr->sa_family == AF_UNSPEC) {
            if (sock->sock_type == SOCK_STREAM && hdl->pal_handle) {
                sock->sock_state = SOCK_CREATED;
                ...
```

Run 1: ```standalone```
ss -unp does NOT show a growing list of sockets

Run 2: with ```gramine-direct```
ss -unp does NOT show a growing list of sockets

Run 3: ```gramine-sgx```
ss -unp shows a steady increase of sockets

Run 4: ```gramine-sgx within a docker container```
ss -unp does NOT show a growing list of sockets

```
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <netdb.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <arpa/inet.h>

int main(int argc, char** argv)
{
    struct addrinfo* result;
    struct addrinfo* res;
    
    if (argc != 3) {
        printf("\n    Usage: %s host delay\n\n", argv[0]);
        exit(1);
    }
    
    unsigned int delay = atoi(argv[2]);
    if (delay > 10) {
        delay = 10;
        printf("delay cap to 10 s.\n");
    }
    
    for (;;) {
        
        int error;
        error = getaddrinfo(argv[1], NULL, NULL, &result);
        if (error != 0) {
            if (error == EAI_SYSTEM) {
                perror("getaddrinfo");
            } else {
                fprintf(stderr, "error in getaddrinfo: %s\n", gai_strerror(error));
            }
            exit(EXIT_FAILURE);
        }
        
        freeaddrinfo(result);
        
        sleep(delay);
        
    }
    
    return 0;
}
```

**Comment by dimakuv** [comment]
@boryspoplawski You discussed recently the issues with "reconnects on sockets". Have you done anything in that area? Do you recognize the memory leak described here? Does the fix make sense to you?

**Comment by boryspoplawski** [comment]
@dimakuv not really, the problem I've encountered was that `PAL_HANDLE` can arbitrarily change, but there is no locking around that.

**Comment by dimakuv** [comment]
I cannot reproduce this problem on my local machine with Ubuntu 18.04 and my `/etc/nsswitch.conf` looking like this:
```
passwd:         compat systemd
group:          compat systemd
shadow:         compat
gshadow:        files

hosts:          files mdns4_minimal [NOTFOUND=return] dns myhostname
networks:       files

protocols:      db files
services:       db files
ethers:         db files
rpc:            db files

netgroup:       nis
```

My `/etc/resolv.conf` contains a single nameserver (a corporate DNS machine).

My `/etc/gai.conf` is empty.

From what I understand, these are the only important files required for `getaddrinfo()` functioning. @xraw Could you share your files? Or explain a bit more on how to reproduce your issue?

**Comment by boryspoplawski** [comment]
I cannot reproduce this issue either. I used `/etc/resolv.conf` with `nameserver 8.8.8.8` and rest of files empty, as well as default ubuntu conf (`/etc` passed through to Gramine, with 8.8.8.8 in resolv.conf)

@xraw any more details on how to reproduce this? From what I've seen it does not reuse sockets, i.e. when I compiled this snipped, it always closed the old socket and created a new one. But that probably depends on glibc version and what not...

Update: I was able to somewhat reproduce it: just created a simple program doing reconnects on UDP socket. `ss` did not show me anything, but I was able to see a growing list of fds in `/proc/[pid]/fd/`.

Fortunately #579 fixes this
